### PR TITLE
Turn item hold effects into an enum

### PR DESCRIPTION
**Pull request by Bassoonian**
Turns hold effects into an enum for easier editing. Also removes a handful of references to functions that no longer exist in `battle_ai_util.h`.

## **Discord contact info**
bassoonian

### CODE REVIEW

File: include/constants/hold_effects.h [hunk 1]
```diff
@@ -1,171 +1,167 @@
...
-#define HOLD_EFFECT_FIRE_POWER     58
-#define HOLD_EFFECT_DRAGON_POWER   59
-#define HOLD_EFFECT_NORMAL_POWER   60
-#define HOLD_EFFECT_UPGRADE        61
-#define HOLD_EFFECT_SHELL_BELL     62
-#define HOLD_EFFECT_LUCKY_PUNCH    63
-#define HOLD_EFFECT_METAL_POWDER   64
-#define HOLD_EFFECT_THICK_CLUB     65
-#define HOLD_EFFECT_LEEK           66
+enum ItemHoldEffects
```
Comment (mrgriffin) [inline include/constants/hold_effects.h:4]: I'd call this `enum ItemHoldEffect`. I think this makes more sense when it comes to variables, e.g. `Foo(enum ItemHoldEffect effect)` takes a single effect, not multiple effects.

I haven't checked, but if the plural naming already has precedent in Expansion I'll try and "correct" them later :)
```suggestion
enum ItemHoldEffect
```

**Comment by AlexOn1ine** [comment]
`ItemHoldEffect` should be used as a variable identifier in functions where possible

**Comment by Bassoonian** [comment]
> `ItemHoldEffect` should be used as a variable identifier in functions where possible

I tried this, but I got buried in `werror=switch` errors because for some reason, supplying a switch with an enum will complain if you don't check _each and every enum_ (or have a default case, I guess)

**Comment by mrgriffin** [comment]
> I tried this, but I got buried in werror=switch errors because for some reason, supplying a switch with an enum will complain if you don't check each and every enum (or have a default case, I guess)

Just to explain the reason: it's so that if you add a new case to your `enum` the compiler will point out all the places that you need to update (unless there's a `default`, as you [afaik] correctly guess).

As a motivating example, consider something like script commands. If we had an `enum BattleScriptCommand { BSC_NOP, ... BSC_VARIOUS, ... }` then the warning is very likely the right thing, because people rarely write `switch (command)` and don't want to handle every kind of command.

**Review by AlexOn1ine — APPROVED** [review]

**Comment by mrgriffin** [comment]
I'll PR https://github.com/mrgriffin/pokeemerald/commit/c24763360220f0877cd7d2bf9bec70529972d0ef which contains usages of `enum ItemHoldEffect` once (if?) https://github.com/pret/pokeemerald/pull/2043 is accepted.

**Comment by AlexOn1ine** [comment]
> I'll PR [mrgriffin@c247633](https://github.com/mrgriffin/pokeemerald/commit/c24763360220f0877cd7d2bf9bec70529972d0ef) which contains usages of `enum ItemHoldEffect` once (if?) [pret#2043](https://github.com/pret/pokeemerald/pull/2043) is accepted.

oops, sorry :sweat_smile:

**Pull request merged**